File: 2021/day2/day2.py
with open('day2.txt') as file:
    steps = tuple([tuple(line.strip().split(' ')) for line in file.readlines()])

# ----------
# | Part 1 |
# ----------

depth = 0
horizontal = 0

for step in steps:
    amount = int(step[1])

    if step[0] == 'forward':
        horizontal += amount
    elif step[0] == 'down':
        depth += amount
    elif step[0] == 'up':
        depth -= amount

print('Part 1:')
print(f'Horizontal Position: {horizontal}')
print(f'Depth: {depth}')
print(f'Result: {horizontal * depth}\n')

# ----------
# | Part 2 |
# ----------

aim = 0
depth = 0
horizontal = 0

for step in steps:
    amount = int(step[1])

    if step[0] == 'forward':
        horizontal += amount
        depth += aim * amount
    elif step[0] == 'down':
        aim += amount
    elif step[0] == 'up':
        aim -= amount

print('Part 2:')
print(f'Aim: {aim}')
print(f'Horizontal Position: {horizontal}')
print(f'Depth: {depth}')
print(f'Result: {horizontal * depth}')

File: 2021/day2/day2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>

int main()
{
    std::string line;
    std::ifstream file("./day2.txt");

    int p1_horizontal = 0;
    int p1_depth = 0;

    int p2_aim = 0;
    int p2_horizontal = 0;
    int p2_depth = 0;

    while (std::getline(file, line))
    {
        std::string direction = line.substr(0, line.find(" "));
        int distance = std::stoi(line.substr(line.find(" ") + 1));

        if (direction == "down")
        {
            p1_depth += distance;
            p2_aim += distance;
        }
        else if (direction == "up")
        {
            p1_depth -= distance;
            p2_aim -= distance;
        }
        else if (direction == "forward")
        {
            p1_horizontal += distance;
            p2_depth += distance * p2_aim;
            p2_horizontal += distance;
        }
    }

    std::cout << "Part 1: " << p1_horizontal * p1_depth << "\n";
    std::cout << "Part 2: " << p2_horizontal * p2_depth << "\n";
}